client/core/trade.go

```diff
@@ -214,9 +214,9 @@ func (t *trackedTrade) accountRefunder() (asset.AccountLocker, bool) {
 	return ar, is
 }
 
-// lockRedemptionFraction locks the specified fraction of the available
+// lockRefundFraction locks the specified fraction of the available
 // refund reserves. Subsequent calls are additive. If a call to
-// lockRedemptionFraction would put the locked reserves > available reserves,
+// lockRefundFraction would put the locked reserves > available reserves,
```

Comment:
Not related to your changes, but I'm having a very hard time understanding why we're going to so much trouble of working with a rational number's numerator and denominator if we just end up doing this:

```
func applyFraction(num, denom, target uint64) uint64 {
	return uint64(math.Round(float64(num) / float64(denom) * float64(target)))
}
```

We're _are_ going to see precision loss and likely the occasional wrong result, but we still have to jump through all these hoops.  What am I missing?

I'm looking at `math/big.Rat` if we want to be rigorous w.r.t. precision.  Thoughts or clarification?

Reply:
```go
func applyFraction(num, denom, target uint64) uint64 {
	frac := big.NewRat(int64(num), int64(denom))
	res, exact := new(big.Rat).Mul(frac, big.NewRat(int64(target), 1)).Float64()
	fmt.Println("**************** frac:", frac)
	fmt.Println("**************** val:", target)
	fmt.Printf("**************** scaled: %v (f) => %d (i) (exact = %v)\n", res, uint64(math.Round(res)), exact)
	simple := uint64(math.Round(float64(num) / float64(denom) * float64(target)))
	fmt.Println("**************** simple:", simple)
	return simple
}
```

Exposes some issues with `TestRedemptionReserves` and `TestRefundReserves`.  In particular trying to divide an order size by a value that the order cannot be split evenly.

Test issues aside, I'm still thinking that we an change the `trackedTrade` fields to a new system whereby for say, redeem reserves, one field is the multiple size and the other field is the number of reserved multiples.  Then the unlock methods just need to specify an integer multiple to unlock, which just decrements the locked integer.

Reply:
Nothing to change with this PR, @JoeGruffins .  Just clarifying

Reply:
I agree. The current locker does deal with "dust" if a remainder slipped through, but there's probably a more strait forward answer. I'll make an issue?

Reply:
> I'll make an issue?

For what?

Reply:
The fraction system is turning out to be brittle, complex, and prone to precision loss and rounding error.  If we continue to have trouble with reserves locking, I'll propose some changes as I described in the last paragraph here: https://github.com/decred/dcrdex/pull/1556#discussion_r846523350
No need to divert effort yet.

Reply:
But as @JoeGruffin's pointed out, your comment 

> Exposes some issues with TestRedemptionReserves and TestRefundReserves. In particular trying to divide an order size by a value that the order cannot be split evenly.

is incorrect, since we deal with rounding errors in `reservesToUnlock`.

There've clearly been some hiccups in implementing refunds, but that doesn't mean the system is brittle or complex.

Not trying to stop you from trying your plan though.

Reply:
I don't mean to harshly criticize the design here; I just believe that it does not handle rounding well and it its use is error prone.

In `applyFraction`, the `math.Round` deals with floating point _precision_ errors, not errors in the input arguments that create an expression where the number cannot be divided evenly even on paper, never mind floating point precision.  The main issue is not when precision results in something like 3999.9999997, and that rounds to 4000 just fine; it's when num and denom result in something like 13333.4 or 666.6666666.  The result will just be wrong because the caller gave unsatisfiable inputs.

I _sort of_ understand the dust handling in `reservesToUnlock`, but why should there be dust.  Also what if the rounding in applyFraction goes the other way an is unlocking one extra gwei instead?  Is that all handled?  I cannot tell.

In the tests, run with the modified code I posted above.  First thing I noticed is this:

```
**************** frac: 2/3
**************** val: 100000
**************** scaled: 66666.66666666667 (f) => 66667 (i) (exact = false)
**************** simple: 66667
```

because of 
```
const reserves = 100_000
qty := lotSize * 10
matchQty := qty * 2 / 3
matchReserves := applyFraction(2, 3, reserves)
```

The math.Round isn't correcting for floating point precision with this; it's just not a whole number _on paper_.  Is this really expected?  I don't get it if so.

Even if this is all somehow "right", there's still little point in keeping num and denom separate because of `math.Round(float64(num) / float64(denom) * float64(target))`, which divides num/denom into a float first anyway.

Reply:
I'm sure a big part of this is just me being slow, but I am confused by how to use the functions and what the arguments mean, and I'm surprised at the acrobatics needed to deal with the reserves.

Reply:
I hear ya. I'll add some notes to explain some of the choices and hopefully save you some time in implementation.

- **Assuming that the redemption and refund reserves are evenly divisible by the number `n` passed to `ReserveNRedemptions`**, then for all orders but market buy orders, everything will work out evenly without rounding.
- The generalization of returns as fractional is to deal with market buy orders, where we can't know at order-time how many matches there will be. We **could** store the `n` value estimate we used to reserve and use that to eliminate rounding for market buy orders.
- I do think that the integer system that you're planning will be mathematically identical to this, just with more data persistence. Consider this. I could, instead of 
 
```Go
returnQty := applyFraction(matchQuantity, orderQuantity, reserves)
```
do
```Go
matchLots := matchQuantity / mktInfo.LotSize
orderLots := orderQuantity / mktInfo.LotSize
returnQty := applyFraction(matchLots, orderLots, reserves)
```

One problem is we're not saving the market info, but let's overlook that for a second.

If we were instead to save **a reserve quantity per lot** (in the spirit of your suggestion, I think), then we might have something like this

```Go
matchLots := matchQuantity / mktInfo.LotSize
returnQty := tracker.reserveQtyPerLot * matchLots
```

Which looks clean, but would require the persistence of more data.

Going back now, internally in `applyFraction`, I do roughly

```Go
return matchQuantity / orderQuantity * reserves
```
or equivalently
```Go
return matchLots / orderLots * reserves
```
Multiplying numerator and denominator by `1 / orderLots`
```Go
return matchLots * (reserves / orderLots)
```

which is exactly where we got with the alternative approach. Same result, more data persistence required. 

> Even if this is all somehow "right", there's still little point in keeping num and denom separate because of `math.Round(float64(num) / float64(denom) * float64(target))`, which divides num/denom into a float first anyway.

I'm not sure I understand. How else could I handle `num` and `denom`?

Reply:
Thanks for describing that all in detail.  I appreciate it.

> I'm not sure I understand. How else could I handle `num` and `denom`?

Because `num` and `denom` only ever get used in `applyFraction` (afaict), it could be: 

```go
// frac is on [0,1], target is an amount in atomic units.
func applyFraction(frac float64, target uint64) uint64 {
	return uint64(math.Round(frac * float64(target)))
}
```

Given the above, unless we got fancy with integer math or big.Rat, I'm not seeing a reason to make the caller worry about espressing the frac as a rational number.

Reply:
I was just trying to avoid math and conversions at the call sites.